### PR TITLE
Hawq 991. Add support for "HAWQ register" that could register tables by using "hawq extract" output.

### PR DESCRIPTION
**Pull request by xunzhang**
TODO:
- [x] document from comment.
- [x] distributed by syntax.
- [x] feature test
- [x] refactor with ao format framework
- [x] extend hawq extract with create table DDL
- [x] partition table support？
    * comment: in current version, partition table is not supported. We will support this soon.

### CODE REVIEW

File: tools/bin/hawqregister [hunk 1]
```diff
@@ -40,119 +38,126 @@ EXECNAME = os.path.split(__file__)[-1]
...
     parser.remove_option('-h')
     parser.add_option('-?', '--help', action='help')
-    parser.add_option('-h', '--host', help="host of the target DB")
-    parser.add_option('-p', '--port', help="port of the target DB", type='int', default=0)
-    parser.add_option('-U', '--user', help="username of the target DB")
-    return parser
+    parser.add_option('-h', '--host', help='host of the target DB')
+    parser.add_option('-p', '--port', help='port of the target DB', type='int', default=0)
+    parser.add_option('-U', '--user', help='username of the target DB')
+    parser.add_option('-d', '--database', default = 'postgres', dest = 'database', help='database name')
```
Comment (ictmalili) [inline tools/bin/hawqregister]: @xunzhang , I think you can put the interface onto JIRA description, e.g. the parameters description. Thanks

**Comment by xunzhang** [comment]
cc @ictmalili @changleicn

**Comment by changleicn** [comment]
LGTM

File: src/test/feature/ManagementTool/test_hawq_register.cpp [hunk 2]
```diff
@@ -199,9 +199,121 @@ TEST_F(TestHawqRegister, TestNotHDFSPath) {
...
+  util.execute("create table nt(i int) with (appendonly=true, orientation=parquet);");
+  EXPECT_EQ(0, Command::getCommandStatus("hawq register -d " + (string) HAWQ_DB + " -f hdfs://localhost:8020/hawq_register_hawq.paq nt"));
+	util.query("select * from nt;", 3);
+	util.execute("insert into nt values(1);");
+	util.query("select * from nt;", 4);
+  util.execute("drop table nt;");
+}
+
+TEST_F(TestHawqRegister, TestUsage1ParquetRandomly2) {
+  SQLUtility util;
```
Comment (ictmalili) [inline src/test/feature/ManagementTool/test_hawq_register.cpp]: @xunzhang What's the difference between these two test cases: TestUsage1ParquetRandomly and TestUsage1ParquetRandomly2?

Comment (xunzhang) [inline src/test/feature/ManagementTool/test_hawq_register.cpp]: @ictmalili yes, I made a mistake here, I will update the commit soon.

File: tools/bin/hawqregister [hunk 3]
```diff
@@ -40,186 +39,195 @@ EXECNAME = os.path.split(__file__)[-1]
...
+    parser.add_option('-c', '--config', dest = 'yml_config', default = '', help='configuration file in YAML format')
+    return parser.parse_args()
+
+
+def option_parser_yml(yml_file):
+    import yaml
+    with open(yml_file, 'r') as f:
+        params = yaml.load(f)
+    if params['FileFormat'] == 'Parquet':
+        offset = params['Parquet_FileLocations']['Files'][0]['path'].rfind('/')
```
Comment (ictmalili) [inline tools/bin/hawqregister]: What is the yml file has corrupted? For example, params['Parquet_FileLocations']['Files'] is null?

File: tools/bin/hawqregister [hunk 4]
```diff
@@ -40,186 +39,195 @@ EXECNAME = os.path.split(__file__)[-1]
...
+        filepath = params['DFS_URL'] + params['Parquet_FileLocations']['Files'][0]['path'][:offset] if len(params['Parquet_FileLocations']['Files']) != 1 else params['DFS_URL'] + params['Parquet_FileLocations']['Files'][0]['path']
```
Comment (ictmalili) [inline tools/bin/hawqregister]: Could we change the line to multiple lines?

File: tools/bin/hawqregister [hunk 5]
```diff
@@ -40,186 +39,195 @@ EXECNAME = os.path.split(__file__)[-1]
...
-            logger.error("target not found in table gp_distribution_policy")
+        conn.commit()
+        if not rows.rowcount:
+            logger.error('Target not found in table gp_distribution_policy.')
             sys.exit(1)
         for row in rows:
-            if row[0] != None:
-                logger.error("Cannot register file(s) to a table which is hash-typed")
+            if row[0]:
+                logger.error('Cannot register file(s) to a table which is hash-typed.')
```
Comment (ictmalili) [inline tools/bin/hawqregister]: I think we can change 'hash-typed' to 'hash distributed'

File: tools/bin/hawqregister [hunk 6]
```diff
@@ -40,186 +39,195 @@ EXECNAME = os.path.split(__file__)[-1]
...
         query = "select attrnums from gp_distribution_policy, pg_class where pg_class.relname = '%s' and pg_class.oid = gp_distribution_policy.localoid;" % tablename
-        dburl = dbconn.DbURL(hostname=options.host, port=options.port, username=options.user, dbname=databasename)
         conn = dbconn.connect(dburl, False)
         rows = dbconn.execSQL(conn, query)
-	conn.commit()
-        if rows.rowcount == 0:
-            logger.error("target not found in table gp_distribution_policy")
+        conn.commit()
+        if not rows.rowcount:
+            logger.error('Target not found in table gp_distribution_policy.')
```
Comment (ictmalili) [inline tools/bin/hawqregister]: Will this thrown out to user? If yes, we need to modify the warning content. What's the meaning of 'Target'?

File: tools/bin/hawqregister [hunk 7]
```diff
@@ -40,186 +39,195 @@ EXECNAME = os.path.split(__file__)[-1]
...
     except DatabaseError, ex:
-        logger.error("Failed to connect to database, this script can only be run when the database is up")
-        logger.error("host = %s, port = %d, user = %s, dbname = %s, query = %s" % (options.host, options.port, options.user, databasename, query))
+        logger.error('Failed to execute query "%s"' % query)
         sys.exit(1)
-
     for row in rows:
         tabledir = row[0].strip() + "/" + str(row[1]) + "/" + str(row[2]) + "/" + str(row[3]) + "/"
-
+        #tabledir = '/'.join([row[0], str(row[1]), str(row[2]), str(row[3]), ''])
```
Comment (ictmalili) [inline tools/bin/hawqregister]: Should we remove this line?

File: tools/bin/hawqregister (hunk 5 above)
Comment (xunzhang) [inline tools/bin/hawqregister]: ok, thanks.

File: tools/bin/hawqregister (hunk 4 above)
Comment (xunzhang) [inline tools/bin/hawqregister]: Sure.

File: tools/bin/hawqregister (hunk 7 above)
Comment (xunzhang) [inline tools/bin/hawqregister]: I will remove L157 and remain L158.

File: tools/bin/hawqregister (hunk 6 above)
Comment (xunzhang) [inline tools/bin/hawqregister]: Yes, it will be delivered to user. I will rename it with `Table`.

File: tools/bin/hawqregister (hunk 3 above)
Comment (xunzhang) [inline tools/bin/hawqregister]: Good point. I will rewrite the code around.

**Comment by ictmalili** [comment]
LGTM

**Comment by xunzhang** [comment]
Merged into master, thanks.

**Pull request closed without merging**